From f8e686fda7c681d1b27e266d3f954d6441aee83a Mon Sep 17 00:00:00 2001
Date: Thu, 13 Feb 2020 19:07:20 +0000
Subject: [PATCH] 8227662: freetype seeks to index at the end of the font data

Summary: <freetype>: freetype seeks to index at the end of the font data
LLT: test/jdk/java/awt/FontMetrics/SpaceAdvance.java
Bug url: https://bugs.openjdk.java.net/browse/JDK-8237381
---
 .../share/native/libfontmanager/freetypeScaler.c   |  2 +-
 test/jdk/java/awt/FontMetrics/SpaceAdvance.java    | 49 ++++++++++++++++++++++
 2 files changed, 50 insertions(+), 1 deletion(-)
 create mode 100644 test/jdk/java/awt/FontMetrics/SpaceAdvance.java

diff --git a/src/java.desktop/share/native/libfontmanager/freetypeScaler.c b/src/java.desktop/share/native/libfontmanager/freetypeScaler.c
index bca003d31..cc8432866 100644
--- a/src/java.desktop/share/native/libfontmanager/freetypeScaler.c
+++ b/src/java.desktop/share/native/libfontmanager/freetypeScaler.c
@@ -161,7 +161,7 @@ static unsigned long ReadTTFontFileFunc(FT_Stream stream,
      */
 
     if (numBytes == 0) {
-        if (offset >= scalerInfo->fileSize) {
+        if (offset > scalerInfo->fileSize) {
             return -1;
         } else {
             return 0;
diff --git a/test/jdk/java/awt/FontMetrics/SpaceAdvance.java b/test/jdk/java/awt/FontMetrics/SpaceAdvance.java
new file mode 100644
index 000000000..e2c7acb6f
--- /dev/null
+++ b/test/jdk/java/awt/FontMetrics/SpaceAdvance.java
@@ -0,0 +1,49 @@
+/*
+ * Copyright (c) 2019, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/*
+ * @test
+ * @bug 8227662
+ */
+
+import java.awt.Font;
+import java.awt.FontMetrics ;
+import java.awt.Graphics2D;
+import java.awt.image.BufferedImage;
+
+public class SpaceAdvance {
+    public static void main(String[] args) throws Exception {
+
+        BufferedImage bi = new BufferedImage(1,1,1);
+        Graphics2D g2d = bi.createGraphics();
+        Font font = new Font(Font.DIALOG, Font.PLAIN, 12);
+        if (!font.canDisplay(' ')) {
+            return;
+        }
+        g2d.setFont(font);
+        FontMetrics fm = g2d.getFontMetrics();
+        if (fm.charWidth(' ') == 0) {
+            throw new RuntimeException("Space has char width of 0");
+        }
+    }
+}
-- 
2.12.3

